homeassistant/components/dwd_weather_warnings/sensor.py

```diff
@@ -166,77 +166,115 @@ def update(self):
 class DwdWeatherWarningsAPI:
     """Get the latest data and update the states."""
 
+    warning_category_dict = {
+        "Minor": 1,
+        "Moderate": 2,
+        "Severe": 3,
+        "Extreme": 4
+    }
+
     def __init__(self, region_name):
         """Initialize the data object."""
-        resource = "{}{}{}?{}".format(
+        self.region_name = region_name
+
+        if not self.region_name:
+            _LOGGER.error('No region configured!')
+            self.available = False
+            return
+
+        if not self.regioncheck():
+            return
+
+        resource = "{}{}{}{}{}{}{}".format(
             'https://',
-            'www.dwd.de',
-            '/DWD/warnungen/warnapp_landkreise/json/warnings.json',
-            'jsonp=loadWarnings'
+            'maps.dwd.de',
```

Comment:
This code needs to be extracted into a third party library that Home Assistant depends on. (and the existing code should never have been allowed in)

Reply:
Thanks for your review!
Shall I just move the DwdWeatherWarningsAPI class to a dwdweatherwarningsapi.py file?

Reply:
Thanks for your review!
Shall I just move the DwdWeatherWarningsAPI class to a dwdweatherwarningsapi.py file?

Reply:
https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services

There's a link in that docs section to a guide about publishing a library to PyPI.

Reply:
Although looking at that guide, it's a bit outdated. Eg twine isn't mentioned. I would recommend this one instead:
https://realpython.com/pypi-publish-python-package/

Reply:
Hi Martin,
thanks for your reply.
Does that also mean that I do not need to use homeassistant rest sensor but instead use some other rest api available via pip?
Also, I am not sure how to handle log entries. Shall I trow exceptions within the third party library and catch it within the sensor? Many sensors use this `_LOGGER = logging.getLogger(__name__)` though I am not sure if this is suitable / good practise for an external API.
Thanks in advance.

Reply:
Yes, don't use the home assistant rest sensor. If there's an existing pypi library for this integration that works, then use that library and you don't need to create a new library.

It's good practice to define a logger per module in libraries:
```py
_LOGGER = logging.getLogger(__name__)
```
Then the app that uses the library can filter logs easily.

It's also good practice for a library to define its own exceptions and raise those as needed. There should be a single base exception class defined in the library that the other library exceptions should inherit from.

There's two ways to go about exceptions in a library, either make sure to catch all exceptions in the library and just log errors. Or raise library specific exceptions.